modules/images/regenerate-existing-images/background-process/class-perflab-background-job.php

```diff
@@ -0,0 +1,449 @@
+	}
+
+	/**
+	 * Retrieves the job data.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @return array|null
+	 */
+	public function data() {
+		// If we have cached data, return it.
+		if ( ! is_null( $this->data ) ) {
+			return $this->data;
+		}
+
+		$this->data = get_term_meta( $this->job_id, self::JOB_DATA_META_KEY, true );
+
+		return $this->data;
+	}
+
+	/**
+	 * Creates a new job in the queue.
+	 *
+	 * Job refers to the term and queue refers to the `background_job` taxonomy.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param string $name Name of job identifier.
+	 * @param array  $data Data for the job.
+	 *
+	 * @return int|WP_Error
+	 */
+	public function create( $name, array $data = array() ) {
+		// @todo Replace taxonomy name with constant.
+		// Create job only when job ID for current instance is zero or term doesn't exist.
+		if ( $this->job_id > 0 || $this->exists() ) {
+			return new WP_Error( __( 'Cannot create the job as it exists already.', 'performance-lab' ) );
+		}
+
+		// Insert the new job in queue.
+		$term_data = wp_insert_term( 'job_' . time(), 'background_job' );
+
+		if ( ! is_wp_error( $term_data ) ) {
+			// Set object properties for instance as soon as job is created.
+			$this->job_id = $term_data['term_id'];
+			$this->name   = sanitize_text_field( $name );
+			$this->data   = $data;
+
+			// Set the queued status for freshly created jobs.
+			$this->set_status( self::JOB_STATUS_QUEUED );
+
+			update_term_meta( $this->job_id, self::JOB_DATA_META_KEY, $this->data );
+			update_term_meta( $this->job_id, self::JOB_NAME_META_KEY, $this->name );
+
+			/**
+			 * Fires when the job has been created successfully.
+			 *
+			 * @since n.e.x.t
+			 *
+			 * @param int    $job_id Job ID.
+			 * @param string $name   Job name.
+			 * @param array  $data   Job data.
+			 */
+			do_action( 'perflab_job_created', $this->job_id, $this->name, $this->data );
+		}
+
+		return $term_data;
+	}
+
+	/**
+	 * Get the batch for current run.
+	 *
+	 * This will return the items to process in the current batch.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @return array
+	 */
+	public function batch() {
+		$items = array();
+
+		/**
+		 * Filters the items for the current batch of job.
+		 *
+		 * By default this will be empty array. Consumer code will need to use this filter in
+		 * order to return the list of items which needs to be processed in current batch.
+		 *
+		 * @since n.e.x.t
+		 *
+		 * @param array  $items  Items for the current batch.
+		 * @param int    $job_id Background job ID.
+		 * @param string $name   Job identifier or name.
+		 * @param array  $data   Job data.
+		 */
+		$items = apply_filters( 'perflab_job_batch_items', $items, $this->job_id, $this->name, $this->data );
+
+		return (array) $items;
+	}
+
+	/**
+	 * Process the batch item.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param mixed $item Batch item for the job.
+	 *
+	 * @return void
+	 */
+	public function process( $item ) {
```

Comment:
@jjgrainger Running the job is performed by Background_Process class itself. Please refer to [run method](https://github.com/WordPress/performance/pull/512/files#diff-9eb4e7423964c54b17e964088ed6431ce365956a5cbe54b2b3c55d4324cd88a7R111) in this PR.

This method (`process`) is mainly there to perform any given task on batch item. Maybe we can rename more appropriately to `process_item`. 

For example: In image regeneration job, item will be attachment post, so this action has been given so that any consumer code can hook there and process item.

Reply:
Thanks @ankitrox 

I think we need to move away from the idea batches in the API and have the majority of logic handled in the custom action created by the consumer. This background process architecture is intended to be used for a variety of use cases that may not need or understand the idea of 'batches'.

For the regeneration job, I would review #487 to see how the consumer job handles batches in its own way, all within the action called by the `Perflab_Background_Process` class. We're using a `WP_Query` and term to determine what attachments need to be processed and handling them all within a single hook. I think separating this process out into 2 hooks could make things more complex than needed.